src/esp/gfx/PbrShader.cpp

```diff
@@ -0,0 +1,484 @@
+    result &= ~Flag::RoughnessTexture;
+    result &= ~Flag::MetallicTexture;
+  } else if (result & Flag::NoneRoughnessMetallicTexture) {
+    result &= ~Flag::RoughnessTexture;
+    result &= ~Flag::MetallicTexture;
+  }
+
+  return result;
+}
+
+PbrShader::PbrShader(Flags originalFlags, unsigned int lightCount)
+    : flags_(generateCorrectFlags(originalFlags)), lightCount_(lightCount) {
+  if (!Cr::Utility::Resource::hasGroup("default-shaders")) {
+    importShaderResources();
+  }
+
+#ifdef MAGNUM_TARGET_WEBGL
+  Mn::GL::Version glVersion = Mn::GL::Version::GLES300;
+#else
+  Mn::GL::Version glVersion = Mn::GL::Version::GL330;
+#endif
+
+  // this is not the file name, but the group name in the config file
+  // see Shaders.conf in the shaders folder
+  const Cr::Utility::Resource rs{"default-shaders"};
+
+  Mn::GL::Shader vert{glVersion, Mn::GL::Shader::Type::Vertex};
+  Mn::GL::Shader frag{glVersion, Mn::GL::Shader::Type::Fragment};
+
+  std::stringstream attributeLocationsStream;
+  attributeLocationsStream << Cr::Utility::formatString(
+      "#define ATTRIBUTE_LOCATION_POSITION {}\n", Position::Location);
+  if (lightCount) {
+    attributeLocationsStream << Cr::Utility::formatString(
+        "#define ATTRIBUTE_LOCATION_NORMAL {}\n", Normal::Location);
+  }
+  if (flags_ & (Flag::NormalTexture | Flag::PrecomputedTangent) && lightCount) {
+    attributeLocationsStream << Cr::Utility::formatString(
+        "#define ATTRIBUTE_LOCATION_TANGENT4 {}\n", Tangent4::Location);
+  }
+  auto isTextured = [&]() {
+    return (flags_ & (Flag::BaseColorTexture | Flag::RoughnessTexture |
+                      Flag::MetallicTexture | Flag::NormalTexture |
+                      Flag::NoneRoughnessMetallicTexture |
+                      Flag::OcclusionRoughnessMetallicTexture));
+  };
+  if (isTextured()) {
+    attributeLocationsStream
+        << Cr::Utility::formatString("#define ATTRIBUTE_LOCATION_TEXCOORD {}\n",
+                                     TextureCoordinates::Location);
+  }
+
+  // Add macros
+  vert.addSource(attributeLocationsStream.str())
+      .addSource(isTextured() ? "#define TEXTURED\n" : "")
+      .addSource(flags_ & Flag::NormalTexture ? "#define NORMAL_TEXTURE\n" : "")
+      .addSource(flags_ & Flag::PrecomputedTangent
+                     ? "#define PRECOMPUTED_TANGENT\n"
+                     : "")
+      .addSource(rs.get("pbr.vert"));
+
+  frag.addSource(attributeLocationsStream.str())
+      .addSource(isTextured() ? "#define TEXTURED\n" : "")
+      .addSource(flags_ & Flag::BaseColorTexture ? "#define BASECOLOR_TEXTURE\n"
+                                                 : "")
+      .addSource(flags_ & Flag::RoughnessTexture ? "#define ROUGHNESS_TEXTURE\n"
+                                                 : "")
+      .addSource(flags_ & Flag::MetallicTexture ? "#define METALLIC_TEXTURE\n"
+                                                : "")
+      .addSource(flags_ & Flag::NoneRoughnessMetallicTexture
```

Comment:
Since I generated the "correct" flags, this piece of code (line 128 - 140) can be as clean as it is now. You can pretend that all textures are "independent", no control logic is needed here.